File: shore-daemon/src/commands/conversation.rs
use serde_json::json;
use shore_protocol::error::ErrorCode;
use shore_protocol::types::{ContentBlock, Message, Role};
use tracing::debug;

use super::{engine_err, CommandContext, CommandResult};
use crate::engine::ConversationEngine;

/// Resolve a message reference to a concrete msg_id.
///
/// Supports: `"last"`, `"latest"`, negative indices (`"-1"` = last),
/// positive 1-based indices (`"3"` = third message), or literal msg_ids (passthrough).
fn resolve_ref(messages: &[Message], reference: &str) -> Result<String, (ErrorCode, String)> {
    if reference == "last" || reference == "latest" {
        return messages
            .last()
            .map(|m| m.msg_id.clone())
            .ok_or_else(|| (ErrorCode::NotFound, "No messages in conversation".into()));
    }

    if let Ok(n) = reference.parse::<i64>() {
        if n == 0 {
            return Err((
                ErrorCode::InvalidRequest,
                "Message index must be non-zero (use 1 for first, -1 for last)".into(),
            ));
        }

        let idx = if n < 0 {
            // -1 = last, -2 = second-to-last, etc.
            (messages.len() as i64) + n
        } else {
            // 1-based positive index
            n - 1
        };

        if idx < 0 || idx as usize >= messages.len() {
            return Err((
                ErrorCode::NotFound,
                format!(
                    "Message index {} out of range (conversation has {} messages)",
                    reference,
                    messages.len()
                ),
            ));
        }

        return Ok(messages[idx as usize].msg_id.clone());
    }

    // Literal msg_id passthrough
    Ok(reference.to_string())
}

/// Get a single message by index or reference.
///
/// Refs resolve against the merged (client-visible) message list.
pub fn get(
    engine: &ConversationEngine,
    _ctx: &CommandContext,
    args: &serde_json::Value,
) -> CommandResult {
    let raw_ref = args.get("ref").and_then(|v| v.as_str()).ok_or_else(|| {
        (
            ErrorCode::InvalidRequest,
            "Missing required argument: ref".into(),
        )
    })?;

    let merged = shore_protocol::merge::merge_tool_loop_messages(engine.messages());
    let msg_id = resolve_ref(&merged, raw_ref)?;
    let msg = merged
        .iter()
        .find(|m| m.msg_id == msg_id)
        .ok_or_else(|| (ErrorCode::NotFound, format!("Message not found: {msg_id}")))?;

    debug!(msg_id = %msg_id, role = ?msg.role, "Message retrieved");
    serde_json::to_value(msg).map_err(|e| (ErrorCode::InternalError, e.to_string()))
}

/// Show conversation history, optionally limited to the last N messages.
///
/// Tool-loop messages are merged into single assistant turns. The `count`
/// parameter applies to the merged (logical) message list.
pub fn log(
    engine: &ConversationEngine,
    _ctx: &CommandContext,
    args: &serde_json::Value,
) -> CommandResult {
    let merged = shore_protocol::merge::merge_tool_loop_messages(engine.messages());

    let count = args
        .get("count")
        .and_then(|v| v.as_u64())
        .map(|c| c as usize);

    // Embed image data so clients can display without filesystem access.
    let mut with_data: Vec<Message> = match count {
        Some(n) => merged.into_iter().rev().take(n).rev().collect(),
        None => merged,
    };
    for msg in &mut with_data {
        crate::handler::embed_image_data(&mut msg.images);
    }

    Ok(json!({ "messages": with_data }))
}

/// Edit a message by ID.
///
/// Refs resolve against the merged (client-visible) message list, then the
/// edit is applied to the raw store by msg_id.
pub fn edit(
    engine: &mut ConversationEngine,
    _ctx: &mut CommandContext,
    args: &serde_json::Value,
) -> CommandResult {
    let raw_ref = args.get("ref").and_then(|v| v.as_str()).ok_or_else(|| {
        (
            ErrorCode::InvalidRequest,
            "Missing required argument: ref".into(),
        )
    })?;

    let content = args
        .get("content")
        .and_then(|v| v.as_str())
        .ok_or_else(|| {
            (
                ErrorCode::InvalidRequest,
                "Missing required argument: content".into(),
            )
        })?;

    let merged = shore_protocol::merge::merge_tool_loop_messages(engine.messages());
    let msg_id = resolve_ref(&merged, raw_ref)?;
    engine.edit_message(&msg_id, content).map_err(engine_err)?;

    debug!(msg_id = %msg_id, content_len = content.len(), "Message edited");
    Ok(json!({ "ref": msg_id, "edited": true }))
}

/// Delete one or more messages by ID.
///
/// `refs` can be a JSON array of strings or a single string.
pub fn delete(
    engine: &mut ConversationEngine,
    _ctx: &mut CommandContext,
    args: &serde_json::Value,
) -> CommandResult {
    let raw_refs: Vec<&str> = if let Some(arr) = args.get("refs").and_then(|v| v.as_array()) {
        arr.iter()
            .map(|v| {
                v.as_str().ok_or_else(|| {
                    (
                        ErrorCode::InvalidRequest,
                        "refs must be an array of strings".into(),
                    )
                })
            })
            .collect::<Result<Vec<_>, _>>()?
    } else if let Some(s) = args.get("refs").and_then(|v| v.as_str()) {
        vec![s]
    } else {
        return Err((
            ErrorCode::InvalidRequest,
            "Missing required argument: refs".into(),
        ));
    };

    // Resolve all refs against merged list, so indices match what users see.
    let merged = shore_protocol::merge::merge_tool_loop_messages(engine.messages());
    let resolved: Vec<String> = raw_refs
        .iter()
        .map(|r| resolve_ref(&merged, r))
        .collect::<Result<Vec<_>, _>>()?;

    let mut deleted = Vec::new();
    for msg_id in &resolved {
        engine.delete_message(msg_id).map_err(engine_err)?;
        deleted.push(msg_id.clone());
    }

    debug!(count = deleted.len(), "Messages deleted");
    Ok(json!({ "deleted": deleted }))
}

/// Inject a system-role instruction into the conversation.
///
/// This allows mid-conversation behavioral correction (e.g. "stop using
/// roleplay actions") without modifying the system prompt or polluting the
/// conversation with user-role meta-instructions.
pub fn inject_system(
    engine: &mut ConversationEngine,
    _ctx: &mut CommandContext,
    args: &serde_json::Value,
) -> CommandResult {
    let text = args.get("text").and_then(|v| v.as_str()).ok_or_else(|| {
        (
            ErrorCode::InvalidRequest,
            "Missing required argument: text".into(),
        )
    })?;

    let msg = Message {
        msg_id: format!("m_{}", uuid::Uuid::new_v4()),
        role: Role::System,
        content: text.to_string(),
        images: vec![],
        content_blocks: vec![ContentBlock::Text {
            text: text.to_string(),
        }],
        alt_index: None,
        alt_count: None,
        timestamp: chrono::Local::now().to_rfc3339(),
    };

    engine.append_message(msg).map_err(engine_err)?;
    debug!(text_len = text.len(), "System message injected");
    Ok(json!({ "injected": true }))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::commands::{CommandContext, SessionTokens};
    use shore_protocol::server_msg::ServerMessage;
    use shore_protocol::types::{Message, Role};
    use tempfile::TempDir;
    use tokio::sync::broadcast;

    fn make_ctx(
        tmp: &TempDir,
    ) -> (
        ConversationEngine,
        CommandContext,
        broadcast::Receiver<ServerMessage>,
    ) {
        let (push_tx, push_rx) = broadcast::channel(16);
        let data_dir = tmp.path().to_path_buf();
        let engine =
            ConversationEngine::new("TestChar".to_string(), data_dir.clone(), push_tx.clone())
                .unwrap();

        let config = shore_config::LoadedConfig::new_for_test(
            shore_config::app::AppConfig::default(),
            shore_config::models::ModelCatalog::default(),
            shore_config::ShoreDirs {
                config: tmp.path().join("config"),
                data: data_dir.clone(),
                runtime: tmp.path().join("runtime"),
            },
        );

        let (_tx, rx) = tokio::sync::watch::channel(());
        let (autonomy, _compaction_rx) = crate::autonomy::manager::AutonomyManager::new(
            Default::default(),
            Default::default(),
            data_dir.clone(),
            rx,
        );

        let ctx = CommandContext {
            config,
            push_tx,
            data_dir: data_dir.clone(),
            active_model: None,
            session_tokens: std::sync::Arc::new(std::sync::Mutex::new(SessionTokens::default())),
            autonomy,
            llm_client: shore_ledger::LedgerClient::new(shore_llm_client::LlmClient::new(), &data_dir.join("ledger.db")).unwrap(),
            diagnostics: std::sync::Arc::new(std::sync::Mutex::new(
                shore_diagnostics::Diagnostics::default(),
            )),
            memory_shell_sessions: std::collections::HashMap::new(),
        };
        (engine, ctx, push_rx)
    }

    fn make_msg(id: &str, role: Role, content: &str) -> Message {
        Message {
            msg_id: id.to_string(),
            role,
            content: content.to_string(),
            images: vec![],
            content_blocks: vec![],
            alt_index: None,
            alt_count: None,
            timestamp: "2026-01-01T00:00:00Z".to_string(),
        }
    }

    #[test]
    fn log_all_messages() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "Hello"))
            .unwrap();
        engine
            .append_message(make_msg("m2", Role::Assistant, "Hi"))
            .unwrap();
        engine
            .append_message(make_msg("m3", Role::User, "How?"))
            .unwrap();

        let result = log(&engine, &ctx, &json!({})).unwrap();
        let msgs = result["messages"].as_array().unwrap();
        assert_eq!(msgs.len(), 3);
    }

    #[test]
    fn log_with_count() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "A"))
            .unwrap();
        engine
            .append_message(make_msg("m2", Role::Assistant, "B"))
            .unwrap();
        engine
            .append_message(make_msg("m3", Role::User, "C"))
            .unwrap();

        let result = log(&engine, &ctx, &json!({"count": 2})).unwrap();
        let msgs = result["messages"].as_array().unwrap();
        assert_eq!(msgs.len(), 2);
        // Should be the last 2 messages.
        assert_eq!(msgs[0]["msg_id"], "m2");
        assert_eq!(msgs[1]["msg_id"], "m3");
    }

    #[test]
    fn log_empty_conversation() {
        let tmp = TempDir::new().unwrap();
        let (engine, ctx, _rx) = make_ctx(&tmp);

        let result = log(&engine, &ctx, &json!({})).unwrap();
        let msgs = result["messages"].as_array().unwrap();
        assert!(msgs.is_empty());
    }

    #[test]
    fn edit_message() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "Original"))
            .unwrap();

        let result = edit(
            &mut engine,
            &mut ctx,
            &json!({"ref": "m1", "content": "Edited"}),
        )
        .unwrap();
        assert_eq!(result["ref"], "m1");
        assert_eq!(result["edited"], true);
        assert_eq!(engine.messages()[0].content, "Edited");
    }

    #[test]
    fn edit_missing_ref() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);

        let result = edit(&mut engine, &mut ctx, &json!({"content": "x"}));
        assert!(result.is_err());
        let (code, _msg) = result.unwrap_err();
        assert_eq!(code, ErrorCode::InvalidRequest);
    }

    #[test]
    fn edit_missing_content() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);

        let result = edit(&mut engine, &mut ctx, &json!({"ref": "m1"}));
        assert!(result.is_err());
        let (code, _msg) = result.unwrap_err();
        assert_eq!(code, ErrorCode::InvalidRequest);
    }

    #[test]
    fn edit_nonexistent_message() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);

        let result = edit(
            &mut engine,
            &mut ctx,
            &json!({"ref": "nope", "content": "x"}),
        );
        assert!(result.is_err());
        let (code, _msg) = result.unwrap_err();
        assert_eq!(code, ErrorCode::NotFound);
    }

    #[test]
    fn edit_triggers_history_push() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, mut rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "Original"))
            .unwrap();
        while rx.try_recv().is_ok() {}

        edit(
            &mut engine,
            &mut ctx,
            &json!({"ref": "m1", "content": "Edited"}),
        )
        .unwrap();

        let msg = rx.try_recv().unwrap();
        assert!(matches!(msg, ServerMessage::History(_)));
    }

    #[test]
    fn delete_messages() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "A"))
            .unwrap();
        engine
            .append_message(make_msg("m2", Role::Assistant, "B"))
            .unwrap();

        let result = delete(&mut engine, &mut ctx, &json!({"refs": ["m1"]})).unwrap();
        assert_eq!(result["deleted"].as_array().unwrap().len(), 1);
        assert_eq!(engine.messages().len(), 1);
        assert_eq!(engine.messages()[0].msg_id, "m2");
    }

    #[test]
    fn delete_single_string_ref() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "A"))
            .unwrap();

        let result = delete(&mut engine, &mut ctx, &json!({"refs": "m1"})).unwrap();
        assert_eq!(result["deleted"].as_array().unwrap().len(), 1);
        assert!(engine.messages().is_empty());
    }

    #[test]
    fn delete_missing_refs() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);

        let result = delete(&mut engine, &mut ctx, &json!({}));
        assert!(result.is_err());
        let (code, _msg) = result.unwrap_err();
        assert_eq!(code, ErrorCode::InvalidRequest);
    }

    #[test]
    fn delete_nonexistent_message() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);

        let result = delete(&mut engine, &mut ctx, &json!({"refs": ["nope"]}));
        assert!(result.is_err());
        let (code, _msg) = result.unwrap_err();
        assert_eq!(code, ErrorCode::NotFound);
    }

    #[test]
    fn delete_triggers_history_push() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, mut rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "A"))
            .unwrap();
        while rx.try_recv().is_ok() {}

        delete(&mut engine, &mut ctx, &json!({"refs": ["m1"]})).unwrap();

        let msg = rx.try_recv().unwrap();
        assert!(matches!(msg, ServerMessage::History(_)));
    }

    // ── resolve_ref tests ───────────────────────────────────────────

    #[test]
    fn resolve_ref_last() {
        let messages = vec![
            make_msg("m1", Role::User, "A"),
            make_msg("m2", Role::Assistant, "B"),
        ];
        assert_eq!(resolve_ref(&messages, "last").unwrap(), "m2");
        assert_eq!(resolve_ref(&messages, "latest").unwrap(), "m2");
    }

    #[test]
    fn resolve_ref_negative_index() {
        let messages = vec![
            make_msg("m1", Role::User, "A"),
            make_msg("m2", Role::Assistant, "B"),
            make_msg("m3", Role::User, "C"),
        ];
        assert_eq!(resolve_ref(&messages, "-1").unwrap(), "m3");
        assert_eq!(resolve_ref(&messages, "-2").unwrap(), "m2");
        assert_eq!(resolve_ref(&messages, "-3").unwrap(), "m1");
    }

    #[test]
    fn resolve_ref_positive_index() {
        let messages = vec![
            make_msg("m1", Role::User, "A"),
            make_msg("m2", Role::Assistant, "B"),
            make_msg("m3", Role::User, "C"),
        ];
        assert_eq!(resolve_ref(&messages, "1").unwrap(), "m1");
        assert_eq!(resolve_ref(&messages, "2").unwrap(), "m2");
        assert_eq!(resolve_ref(&messages, "3").unwrap(), "m3");
    }

    #[test]
    fn resolve_ref_literal_passthrough() {
        let messages = vec![make_msg("m_abc123", Role::User, "A")];
        assert_eq!(resolve_ref(&messages, "m_abc123").unwrap(), "m_abc123");
    }

    #[test]
    fn resolve_ref_zero_is_error() {
        let messages = vec![make_msg("m1", Role::User, "A")];
        let (code, _msg) = resolve_ref(&messages, "0").unwrap_err();
        assert_eq!(code, ErrorCode::InvalidRequest);
    }

    #[test]
    fn resolve_ref_out_of_range() {
        let messages = vec![make_msg("m1", Role::User, "A")];
        let (code, _msg) = resolve_ref(&messages, "99").unwrap_err();
        assert_eq!(code, ErrorCode::NotFound);
        let (code, _msg) = resolve_ref(&messages, "-99").unwrap_err();
        assert_eq!(code, ErrorCode::NotFound);
    }

    #[test]
    fn resolve_ref_empty_conversation() {
        let messages: Vec<Message> = vec![];
        let (code, _msg) = resolve_ref(&messages, "last").unwrap_err();
        assert_eq!(code, ErrorCode::NotFound);
        let (code, _msg) = resolve_ref(&messages, "-1").unwrap_err();
        assert_eq!(code, ErrorCode::NotFound);
    }

    #[test]
    fn edit_by_relative_ref_last() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "First"))
            .unwrap();
        engine
            .append_message(make_msg("m2", Role::Assistant, "Second"))
            .unwrap();

        let result = edit(
            &mut engine,
            &mut ctx,
            &json!({"ref": "last", "content": "Edited"}),
        )
        .unwrap();
        assert_eq!(result["ref"], "m2");
        assert_eq!(engine.messages()[1].content, "Edited");
    }

    #[test]
    fn edit_by_negative_index() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "First"))
            .unwrap();
        engine
            .append_message(make_msg("m2", Role::Assistant, "Second"))
            .unwrap();

        let result = edit(
            &mut engine,
            &mut ctx,
            &json!({"ref": "-1", "content": "Edited"}),
        )
        .unwrap();
        assert_eq!(result["ref"], "m2");
        assert_eq!(engine.messages()[1].content, "Edited");
    }

    #[test]
    fn edit_by_positive_index() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "First"))
            .unwrap();
        engine
            .append_message(make_msg("m2", Role::Assistant, "Second"))
            .unwrap();

        let result = edit(
            &mut engine,
            &mut ctx,
            &json!({"ref": "1", "content": "Edited"}),
        )
        .unwrap();
        assert_eq!(result["ref"], "m1");
        assert_eq!(engine.messages()[0].content, "Edited");
    }

    #[test]
    fn delete_by_relative_ref() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);
        engine
            .append_message(make_msg("m1", Role::User, "A"))
            .unwrap();
        engine
            .append_message(make_msg("m2", Role::Assistant, "B"))
            .unwrap();

        let result = delete(&mut engine, &mut ctx, &json!({"refs": "last"})).unwrap();
        assert_eq!(result["deleted"].as_array().unwrap()[0], "m2");
        assert_eq!(engine.messages().len(), 1);
        assert_eq!(engine.messages()[0].msg_id, "m1");
    }

    #[test]
    fn test_inject_system_appends_message() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);

        let result = inject_system(
            &mut engine,
            &mut ctx,
            &json!({"text": "Stop using actions"}),
        )
        .unwrap();
        assert_eq!(result["injected"], true);

        assert_eq!(engine.messages().len(), 1);
        let msg = &engine.messages()[0];
        assert_eq!(msg.role, Role::System);
        assert_eq!(msg.content, "Stop using actions");
        assert!(msg.msg_id.starts_with("m_"));
    }

    #[test]
    fn test_inject_system_missing_text() {
        let tmp = TempDir::new().unwrap();
        let (mut engine, mut ctx, _rx) = make_ctx(&tmp);

        let result = inject_system(&mut engine, &mut ctx, &json!({}));
        assert!(result.is_err());
        let (code, msg) = result.unwrap_err();
        assert_eq!(code, ErrorCode::InvalidRequest);
        assert!(msg.contains("text"));
    }
}

File: backend/daemon-ts/src/commands/conversation.ts
import { mergeToolLoopMessages } from "../engine/merge.ts";
import type { ConversationEngine } from "../engine/engine.ts";
import type { Message, Role } from "../engine/types.ts";
import {
  asArgs,
  CommandError,
  mapUnknownError,
  requiredString,
} from "./types.ts";

const DEFAULT_LOG_TURNS = 64;

export function getMessage(engine: ConversationEngine, rawArgs: unknown): unknown {
  const args = asArgs(rawArgs);
  const rawRef = requiredString(args, "ref");
  const role = roleFilter(args);
  let merged = mergeToolLoopMessages(engine.messages());
  if (role !== undefined) merged = merged.filter((msg) => msg.role === role);
  const msgId = resolveRef(merged, rawRef);
  const msg = merged.find((m) => m.msg_id === msgId);
  if (msg === undefined) throw new CommandError("not_found", `Message not found: ${msgId}`);
  return msg;
}

export function log(engine: ConversationEngine, rawArgs: unknown): Record<string, unknown> {
  const args = asArgs(rawArgs);
  const { messages, active_start } = engine.displayHistory();
  const end = messages.length;
  const start = pageStartByArgs(messages, end, args);
  const role = roleFilter(args);
  return historyPagePayload(messages, active_start, start, end, role);
}

export function historyPage(engine: ConversationEngine, rawArgs: unknown): Record<string, unknown> {
  const args = asArgs(rawArgs);
  const { messages, active_start } = engine.displayHistory();
  const end = resolveHistoryBefore(args, active_start, messages.length);
  const start = pageStartByArgs(messages, end, args);
  const role = roleFilter(args);
  return historyPagePayload(messages, active_start, start, end, role);
}

export async function editMessage(
  engine: ConversationEngine,
  rawArgs: unknown,
): Promise<Record<string, unknown>> {
  const args = asArgs(rawArgs);
  const rawRef = requiredString(args, "ref");
  const content = requiredString(args, "content");
  const msgId = resolveRef(mergeToolLoopMessages(engine.messages()), rawRef);
  try {
    await engine.editMessage(msgId, content);
  } catch (e) {
    mapUnknownError(e);
  }
  return { ref: msgId, edited: true };
}

export async function deleteMessages(
  engine: ConversationEngine,
  rawArgs: unknown,
): Promise<Record<string, unknown>> {
  const args = asArgs(rawArgs);
  const refs = parseRefs(args);
  const merged = mergeToolLoopMessages(engine.messages());
  const resolved = refs.map((ref) => resolveRef(merged, ref));
  const deleted: string[] = [];
  for (const msgId of resolved) {
    try {
      await engine.deleteMessage(msgId);
    } catch (e) {
      mapUnknownError(e);
    }
    deleted.push(msgId);
  }
  return { deleted };
}

export function listAlternatives(
  engine: ConversationEngine,
  rawArgs: unknown,
): Record<string, unknown> {
  const args = asArgs(rawArgs);
  const rawRef = typeof args["ref"] === "string" ? args["ref"] : undefined;
  try {
    return engine.listAlternatives(rawRef) as unknown as Record<string, unknown>;
  } catch (e) {
    mapUnknownError(e);
  }
}

export async function selectAlternative(
  engine: ConversationEngine,
  rawArgs: unknown,
): Promise<Record<string, unknown>> {
  const args = asArgs(rawArgs);
  const merged = mergeToolLoopMessages(engine.messages());
  const msgId = resolveAssistantRef(merged, typeof args["ref"] === "string" ? args["ref"] : undefined);
  const msg = merged.find((m) => m.msg_id === msgId);
  if (msg === undefined) throw new CommandError("not_found", `Message not found: ${msgId}`);
  const altCount = msg.alternatives?.length ?? 0;
  if (altCount === 0) {
    throw new CommandError("invalid_request", `message ${msgId} has no alternate responses`);
  }
  const current = Math.min(msg.alt_index ?? 0, Math.max(0, altCount - 1));
  const target = resolveAltTarget(args, current, altCount);
  try {
    const selection = await engine.selectAlt(msgId, target);
    return {
      ref: selection.msg_id,
      alt_index: selection.alt_index,
      position: selection.alt_index + 1,
      alt_count: selection.alt_count,
      content: selection.content,
    };
  } catch (e) {
    mapUnknownError(e);
  }
}

export async function injectSystem(
  engine: ConversationEngine,
  rawArgs: unknown,
  now: () => string,
): Promise<Record<string, unknown>> {
  const args = asArgs(rawArgs);
  const text = requiredString(args, "text");
  await engine.appendMessage({
    msg_id: `m_${crypto.randomUUID()}`,
    role: "system",
    content: text,
    images: [],
    content_blocks: [{ type: "text", text }],
    timestamp: now(),
  });
  return { injected: true };
}

function parseRefs(args: Record<string, unknown>): string[] {
  const refs = args["refs"];
  if (Array.isArray(refs)) {
    return refs.map((value) => {
      if (typeof value !== "string") {
        throw new CommandError("invalid_request", "refs must be an array of strings");
      }
      return value;
    });
  }
  if (typeof refs === "string") return [refs];
  throw new CommandError("invalid_request", "Missing required argument: refs");
}

function historyPagePayload(
  messages: Message[],
  globalActiveStart: number,
  startIn: number,
  endIn: number,
  role: Role | undefined,
): Record<string, unknown> {
  const start = Math.min(startIn, messages.length);
  const end = Math.max(start, Math.min(endIn, messages.length));
  const page = messages.slice(start, end).filter((msg) => role === undefined || msg.role === role);
  const archivedEnd = Math.max(start, Math.min(globalActiveStart, end));
  const activePageStart = messages
    .slice(start, archivedEnd)
    .filter((msg) => role === undefined || msg.role === role)
    .length;
  const totalTurns = countUserTurns(messages);
  return {
    messages: page,
    active_start: activePageStart,
    cursor: start,
    next_before: start,
    has_more_before: start > 0,
    global_active_start: globalActiveStart,
    total_messages: totalTurns,
    total_turns: totalTurns,
  };
}

function resolveHistoryBefore(
  args: Record<string, unknown>,
  activeStart: number,
  total: number,
): number {
  const before = args["before"];
  if (before === undefined) return total;
  if (before === "active") return activeStart;
  if (typeof before === "number" && Number.isInteger(before) && before >= 0) {
    return Math.min(before, total);
  }
  throw new CommandError("invalid_request", 'before must be "active" or a message cursor');
}

function pageStartByArgs(
  messages: Message[],
  end: number,
  args: Record<string, unknown>,
): number {
  if (typeof args["turns"] === "number" && Number.isFinite(args["turns"])) {
    return pageStartByTurns(messages, end, Math.max(0, Math.floor(args["turns"])));
  }
  if (typeof args["count"] === "number" && Number.isFinite(args["count"])) {
    return Math.max(0, end - Math.max(0, Math.floor(args["count"])));
  }
  return pageStartByTurns(messages, end, DEFAULT_LOG_TURNS);
}

function pageStartByTurns(messages: Message[], endIn: number, turns: number): number {
  const end = Math.min(endIn, messages.length);
  if (turns === 0) return end;
  let seen = 0;
  for (let idx = end - 1; idx >= 0; idx--) {
    if (messages[idx]?.role === "user") {
      seen += 1;
      if (seen >= turns) return idx;
    }
  }
  return 0;
}

function roleFilter(args: Record<string, unknown>): Role | undefined {
  const role = args["role"];
  if (role === undefined) return undefined;
  if (role === "user" || role === "assistant" || role === "system") return role;
  throw new CommandError("invalid_request", "role must be one of user, assistant, or system");
}

function countUserTurns(messages: Message[]): number {
  return messages.filter((msg) => msg.role === "user" && !isToolResultOnly(msg)).length;
}

function isToolResultOnly(msg: Message): boolean {
  return (
    msg.content_blocks.length > 0
    && msg.content_blocks.every((block) => block.type === "tool_result")
  );
}

function resolveAssistantRef(messages: Message[], reference: string | undefined): string {
  if (reference === undefined || reference === "last" || reference === "latest") {
    const found = [...messages].reverse().find((msg) => msg.role === "assistant");
    if (found === undefined) {
      throw new CommandError("not_found", "No assistant messages in conversation");
    }
    return found.msg_id;
  }
  const msgId = resolveRef(messages, reference);
  const msg = messages.find((m) => m.msg_id === msgId);
  if (msg === undefined) throw new CommandError("not_found", `Message not found: ${msgId}`);
  if (msg.role !== "assistant") {
    throw new CommandError(
      "invalid_request",
      "Alternate response selection only applies to assistant messages",
    );
  }
  return msgId;
}

function resolveRef(messages: Message[], reference: string): string {
  if (reference === "last" || reference === "latest") {
    const last = messages.at(-1);
    if (last === undefined) throw new CommandError("not_found", "No messages in conversation");
    return last.msg_id;
  }
  if (/^-?\d+$/.test(reference)) {
    const n = Number.parseInt(reference, 10);
    if (n === 0) {
      throw new CommandError(
        "invalid_request",
        "Message index must be non-zero (use 1 for first, -1 for last)",
      );
    }
    const idx = n < 0 ? messages.length + n : n - 1;
    if (idx < 0 || idx >= messages.length) {
      throw new CommandError(
        "not_found",
        `Message index ${reference} out of range (conversation has ${messages.length} messages)`,
      );
    }
    return messages[idx]!.msg_id;
  }
  return reference;
}

function resolveAltTarget(
  args: Record<string, unknown>,
  current: number,
  count: number,
): number {
  if (typeof args["index"] === "number" && Number.isInteger(args["index"])) {
    const index = args["index"];
    if (index < 0 || index >= count) {
      throw new CommandError(
        "invalid_request",
        `alternate index ${index + 1} out of range (message has ${count} alternate response(s))`,
      );
    }
    return index;
  }
  if (typeof args["position"] === "number" && Number.isInteger(args["position"])) {
    const position = args["position"];
    if (position <= 0 || position > count) {
      throw new CommandError(
        "invalid_request",
        `alternate position ${position} out of range (message has ${count} alternate response(s))`,
      );
    }
    return position - 1;
  }
  const direction = typeof args["direction"] === "string" ? args["direction"] : "next";
  switch (direction) {
    case "prev":
    case "previous":
      return Math.max(0, current - 1);
    case "next":
      return Math.min(count - 1, current + 1);
    case "first":
      return 0;
    case "last":
      return count - 1;
    default:
      throw new CommandError("invalid_request", `unknown alt direction: ${direction}`);
  }
}
